File: scripts/validate_mermaid.py
#!/usr/bin/env python3
"""
Mermaid Diagram Validator for cv-projects repository.
Checks all ```mermaid blocks across markdown files.
Invokes the official JS Mermaid parser via bun/node if available,
and performs static structural heuristics to ensure diagrams do not fail.
"""

import re
import shutil
import subprocess
import sys
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parent.parent

def check_mermaid_static(repo_root: Path) -> list[dict]:
    """
    Static heuristic validation for common Mermaid syntax pitfalls:
    - Subgraph titles with spaces/parentheses not quoted or lacking an ID
    - Node labels with special characters (parentheses, slashes, arrows) not enclosed in quotes
    - Nested unquoted brackets like [Teacher [CLS] Token Logits]
    """
    errors = []
    mermaid_re = re.compile(r'```mermaid\s*\n(.*?)```', re.DOTALL)
    
    for md_path in repo_root.glob("**/*.md"):
        if any(p in md_path.parts for p in ['.git', 'node_modules', '.venv', '.obsidian']):
            continue
            
        content = md_path.read_text(encoding="utf-8")
        matches = mermaid_re.findall(content)
        
        for idx, diagram in enumerate(matches, 1):
            lines = diagram.strip().split("\n")
            for line_no, line in enumerate(lines, 1):
                sline = line.strip()
                if not sline or sline.startswith("%%"):
                    continue
                
                # Check for bad subgraph syntax
                m_sub = re.match(r'^subgraph\s+([^\[\]\n]+)$', sline)
                if m_sub:
                    title = m_sub.group(1).strip()
                    if ("(" in title or ")" in title or " " in title) and not ('[' in title and ']' in title):
                        errors.append({
                            "file": str(md_path.relative_to(repo_root)),
                            "diagram": idx,
                            "line": line_no,
                            "reason": f"Unquoted subgraph header with spaces/parentheses: '{sline}'",
                        })
                        
                # Check for unquoted arrows inside node definitions
                if re.search(r'\[[^"\'\]]*->[^"\'\]]*\]', sline):
                    errors.append({
                        "file": str(md_path.relative_to(repo_root)),
                        "diagram": idx,
                        "line": line_no,
                        "reason": f"Unquoted '->' inside node brackets: '{sline}'",
                    })

    return errors

def main() -> int:
    bun_path = shutil.which("bun")
    js_script = REPO_ROOT / "scripts" / "validate_mermaid.js"
    
    if bun_path and js_script.exists():
        try:
            res = subprocess.run([bun_path, str(js_script)], cwd=str(REPO_ROOT), capture_output=True, text=True)
            if res.returncode == 0:
                print(res.stdout.strip())
                return 0
            elif "Cannot find package" not in res.stderr and "Cannot find module" not in res.stderr:
                print(res.stdout.strip())
                print(res.stderr.strip(), file=sys.stderr)
                return res.returncode
        except Exception:
            pass
    # 2. Fallback to Python static checks
    static_errors = check_mermaid_static(REPO_ROOT)
    if static_errors:
        print(f"[-] Static Mermaid validation failed ({len(static_errors)} error(s)):")
        for err in static_errors:
            print(f"    - {err['file']} (dia #{err['diagram']}, line {err['line']}): {err['reason']}")
        return 1

    print("[+] Static Mermaid syntax validation passed.")
    return 0

if __name__ == "__main__":
    sys.exit(main())

File: scripts/validate_mermaid.js
#!/usr/bin/env bun
/**
 * Strict Mermaid Diagram Validator for cv-projects repository.
 * Parses all ```mermaid blocks across markdown files using official Mermaid JS parser.
 * Exits with code 0 on success, or 1 with diagnostic details on any parse error.
 */

import fs from 'fs';
import path from 'path';
import { JSDOM } from 'jsdom';

// Initialize headless DOM environment required by Mermaid in non-browser runtimes
const dom = new JSDOM('<!DOCTYPE html><html><body></body></html>');
globalThis.window = dom.window;
globalThis.document = dom.window.document;
globalThis.DOMPurify = (await import('dompurify')).default(dom.window);

const mermaid = (await import('mermaid')).default;
mermaid.initialize({ startOnLoad: false, suppressErrorRendering: true });

function findMarkdownFiles(dir) {
  let results = [];
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    const fullPath = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      if (
        entry.name !== '.git' &&
        entry.name !== 'node_modules' &&
        entry.name !== '.obsidian' &&
        entry.name !== '.venv'
      ) {
        results = results.concat(findMarkdownFiles(fullPath));
      }
    } else if (entry.isFile() && entry.name.endsWith('.md')) {
      results.push(fullPath);
    }
  }
  return results;
}

async function validateAllMermaid() {
  const files = findMarkdownFiles('.');
  const mermaidRegex = /^```mermaid\s*\n([\s\S]*?)```/gm;

  let totalDiagrams = 0;
  let passedDiagrams = 0;
  let errors = [];

  for (const file of files) {
    const content = fs.readFileSync(file, 'utf-8');
    let match;
    let index = 0;
    while ((match = mermaidRegex.exec(content)) !== null) {
      totalDiagrams++;
      index++;
      const code = match[1].trim();
      try {
        await mermaid.parse(code);
        passedDiagrams++;
      } catch (err) {
        errors.push({
          file,
          index,
          error: err.message?.split('\n')[0] || String(err),
          snippet: code.split('\n').slice(0, 5).join('\n')
        });
      }
    }
  }

  console.log(`[+] Scanned ${files.length} markdown files, found ${totalDiagrams} Mermaid diagrams.`);

  if (errors.length > 0) {
    console.error(`[-] Validation FAILED: ${errors.length} diagram(s) failed parsing:`);
    for (const e of errors) {
      console.error(`    - ${e.file} (diagram #${e.index}): ${e.error}`);
      console.error(`      Snippet: ${e.snippet.replace(/\n/g, ' ')}`);
    }
    process.exit(1);
  } else {
    console.log(`[+] Validation PASSED: All ${totalDiagrams} Mermaid diagrams parsed successfully.`);
    process.exit(0);
  }
}

await validateAllMermaid();
